examples/rosettacode/rock_paper_scissor.nit

```diff
@@ -0,0 +1,56 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+# This program is public domain
+#
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+# author: Istvan szalai 22/05/2015
+
+# Return a value to dertermine who win the match.
+# Return 0 for lose, 1 for win, 2 for equality.
+fun win_or_lose(player, computer: String): Int do
+	if player == computer then return 2
+	if player == "rock" and computer == "scissors" then return 1
+	if player == "rock" and computer == "paper" then return 0
+	if player == "paper" and computer == "rock" then return 1
+	if player == "paper" and computer == "scissors" then return 0
+	if player == "scissors" and computer == "paper" then return 1
+	return 0
+end
+
+# Return a string with about the match winner or looser.
+fun result_of_the_match_message(number: Int): String do
+	if number == 2 then return "It's a tie!"
+	if number == 1 then return "You win!"
+	return "You lose!"
+end
+
+# Return a String for the player about what the computer player did.
+fun computer_message(number: String): String do
+	if number == "rock" then return "The computer played rock"
+	if number == "scissors" then return "The computer played scissors"
+	return "The computer played paper"
+end
+
+# Return a random value for the computer player input.
+fun computer_player : String do
+	var tab = ["rock", "scissors", "paper"]
+	return tab.rand
+end
+
+# Game loop.
+var x = ""
+var test = false
+loop
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
+		if x == "paper" or x == "rock" or x == "scissors" then test = true
```

Comment:
Unfortunately, immutability will not helps you here because there is two issues: (i) what bad things could happen if I reuse the array but it has escaped? Immutability will helps a little because it ensure that the escaped array win not change (but the element may change, you need deep immutability in fact). (ii) if we reuse the array, how confident are we that the initial data make sense in the current context.

To illustrate (ii) lets use this example

~~~nit
fun is_one_or_the_other(object, one, other: Object): Bool do
  return [one, other].has(object)
end

print is_one_or_the_other("a", "a", "c") # true
print is_one_or_the_other("a", "b", "c") # false
~~~

If you assume that the literal array is reused then this means that its content ["a", "c"] is also reused, thus the second call of the method will not have the expected result.

To summarize, in order to reuse a literal array you need that
 * the array is deep-immutable
 * the value of the elements are independent of the state of the computer (parameters, `sys`, I/O, or anything that may depend on `sys`, I/O, or other things)
 
Both points will be very hard to make them fit with the Nit philosophy, so unlikely to be specific or implemented.

Note that if you can control the mutation, the escaping and the refinements, then you can use the infamous and undocumented `once` to reuse the literal array if you want.